### PR TITLE
Refactor number system

### PR DESCRIPTION
**Pull request by Christewart**
This pull request implements the following 
- Native `UInt32`, `UInt64`, `Int32`, and `Int64` data types to align our code base closer to the bitcoin core protocol
- Integrates ScalaCheck into our codebase, which allows for property based testing giving us more certainty on the correctness of consensus critical parts of our codebase

### CODE REVIEW

File: src/main/scala/org/bitcoins/core/gen/ScriptGenerators.scala [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  } yield P2SHScriptPubKey(randomScriptPubKey)
+
+
+  def emptyScriptPubKey = p2pkScriptPubKey.map(_ => EmptyScriptPubKey)
+
+  private def pickRandomNonP2SHScriptPubKey : Gen[ScriptPubKey] = {
+    val randomNum = (scala.util.Random.nextInt() % 3).abs
+    if (randomNum == 0) p2pkScriptPubKey
+    else if (randomNum == 1) p2pkhScriptPubKey
+/*    else if (randomNum == 2) emptyScriptPubKey*/
```
Comment (Christewart) [inline src/main/scala/org/bitcoins/core/gen/ScriptGenerators.scala]: Should this be in the the generator? Come back and review this

File: src/main/scala/org/bitcoins/core/protocol/CompactSizeUInt.scala [hunk 2]
```diff
@@ -22,22 +28,37 @@ trait CompactSizeUInt {
 
   def hex = size match {
     case 1 => if (num.toHexString.size == 1) "0" + num.toHexString else num.toHexString
-    case 3 =>
-      "fd" + BitcoinSUtil.encodeHex(BitcoinSUtil.toByteSeq(num).reverse)
-    case 5 => "fe" + BitcoinSUtil.encodeHex(BitcoinSUtil.toByteSeq(num).reverse)
-    case _ => "ff" + BitcoinSUtil.encodeHex(BitcoinSUtil.toByteSeq(num).reverse)
+    case 3 => "fd" + ScriptNumberUtil.longToHex(num)
```
Comment (Christewart) [inline src/main/scala/org/bitcoins/core/protocol/CompactSizeUInt.scala]: Don't think these should be calling `ScriptNumberUtil`, if I remember correctly the numbers are simply unsigned integers which could go through `UInt32` or `UInt64` possibly

**Comment by TomMcCabe** [comment]
ACK.

**Comment by Christewart** [comment]
ACK.

**Pull request merged**